### ISSUE TITLE
How to modify weights in TFLite model and check effect on activation layer output, set_tensor helps only on input layers and not intermediate tensors.

### ISSUE THREAD

**Issue by manp-git** [tensorflow/tensorflow#48879]
How to modify weights in TFLite model and check effect on activation layer output for experimentation purpose, set_tensor helps reapply things only on input layers and not intermediate tensors.
I am ok doing in offline by extracting output of previous layer and convolving it with the extracted weights and biases of layer in consideration. 
Please share if there is a command for that. Else how can I exactly replicate the convolution and activation process of that particular Resnet50 layer outside the model using python etc.
Using standard Resnet50- keras model, quantised by TFLite converter.

Targetting conv2_block1_2 for experimentation; able to extract output of previous conv block which is input to this (conv2_block1_1) along with weights (ReadVariableOp) and bias of conv2_block1_2 from the TFLite model.

Tensorflow 2.0

**Comment by tilakrayal** [comment]
@manp-git ,

In order to reproduce the issue reported here, could you please provide  the complete code and the dataset  or colab link you are using. Thanks!

**Comment by abattery** [comment]
To modify the weight of the specific graph nodes, you need to edit the flatbuffer library based on the TFLite schema directly.

The TensorFlow Lite model file is based on the flatbuffer file format.

**Comment by manp-git** [comment]
> To modify the weight of the specific graph nodes, you need to edit the flatbuffer library based on the TFLite schema directly.
> 
> The TensorFlow Lite model file is based on the flatbuffer file format.

Thanks @abattery. 
So I found the following links related to what you mentioned:-
https://www.tensorflow.org/lite/convert/metadata
https://www.tensorflow.org/lite/api_docs/cc/class/tflite/flat-buffer-model#classtflite_1_1_flat_buffer_model_1a658e48aadda1eafb55fc7deb0434da14
https://github.com/tensorflow/tflite-support/blob/master/tensorflow_lite_support/metadata/metadata_schema.fbs
https://tensorflow.google.cn/lite/api_docs/cc/class/tflite/flat-buffer-model?hl=it
https://analyticsindiamag.com/my-first-cnn-project-emotion-detection-using-convolutional-neural-network-with-tpu/

Is there a template available which I can use to access the flatbuffer file and edit the weights and then execute the model to check effect of changed weights on the relu output of that particular node?

Does it have anything to do with JSON as i searched?

Sorry very new to compuster science and python.

**Comment by thaink** [comment]
@manp-git You can follow https://google.github.io/flatbuffers/flatbuffers_guide_tutorial.html
There is a section about mutating Flatbuffers. Our schema is https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/schema/schema.fbs

**Comment by manp-git** [comment]
@thaink , thanks,

As per what I understand from the above links is that:

I nee to install flatbuffers from Anaconda by doing:
conda install -c conda-forge flatbuffers
Then run the flatc command on the schema and model:
flatc -t schema.fbs -- model.tflite

Then can call the table elements in schema using monster
Then can edit using mutate (python support not available/ other languages to try)

And then may be recompile to generate the new model.

But i can find where the weights (64x3x3x64) will be in the schema.fbs,
Is there a template to get this done?
Or some simpler way?

Very new to Computer science, never heard of these.

Zip file of the resnet-50 tflite model attached; Has those layers only that I pulled out for observability.
[tflite_model_int8_w_conv_6.zip](https://github.com/tensorflow/tensorflow/files/6433155/tflite_model_int8_w_conv_6.zip)

Targetting conv2_block1_2 for experimentation; successfully able to extract output of previous conv block which is input to this (conv2_block1_1) along with weights (ReadVariableOp) and bias of conv2_block1_2 from the TFLite model.

**Comment by manp-git** [comment]
Hi @thaink @abattery 
This what I have done to replicate the resnet tflite model functionality outside the model, this would help me get more flexibility over my intended project:-

Following tensors extracted from TFLite Resnet50 model for experimentation on convolution.
in1.shape = (1,56,56,64)    (output from previous relu)
weight_kernel.shape = (64,3,3,64)       (weight of conv_relu under consideration)


in1 -> [1, height, width, channels] -> (1,56,56,64)
weight_kernel -> [height, width, in_channels, out_channels] -> (3,3,64,64)

Apply:-
image_conv = tensorflow.nn.conv2d(in1, weight_kernel, strides=(1,1), padding='SAME', dilations=None)

Now this gives me shape of output of conv, same as input shape -> (1,56,56,64)



Over this I have added the biases that were extracted from the resnet model , my_bias.shape (64,) - as shown in python

this is done by simply doing image_conv + my_bias

the output dimension remains the same - (1,56,56,64)


Now I need to apply Relu activation from my side and compare with the existing relu output

The output from conv+bias is not visible from tflite model so cant compare.

Relu output is also different; and is mentioned as Fused activation function rather than just activation

My ques is how can I replicate this computation outside the model, as a second option.

![image](https://user-images.githubusercontent.com/82860513/117398438-01dae580-af1c-11eb-8020-fb9ac0cf94f3.png)

**Comment by manp-git** [comment]
@tilakrayal the #48933 was for me to resolve the conv issue, which I had posted as a query, then found a workaround myself.

**Comment by thaink** [comment]
@manp-git I think the C++ example is prettty clear:
````
auto monster = GetMutableMonster(buffer_pointer);  // non-const
monster->mutate_hp(10);                      // Set the table `hp` field.
monster->mutable_pos()->mutate_z(4);         // Set struct field.
monster->mutable_inventory()->Mutate(0, 1);  // Set vector element.
````
Here, replace buffer_pointer by the content of the tflite after reading it to memory.
And replace mutate_* and mutable_* function by what you are planning to do.

**Comment by manp-git** [comment]
> Hi @thaink @abattery 
> This what I have done to replicate the resnet tflite model functionality outside the model, this would help me get more flexibility over my intended project:-
> 
> Following tensors extracted from TFLite Resnet50 model for experimentation on convolution.
> in1.shape = (1,56,56,64)    (output from previous relu)
> weight_kernel.shape = (64,3,3,64)       (weight of conv_relu under consideration)
> 
> 
> in1 -> [1, height, width, channels] -> (1,56,56,64)
> weight_kernel -> [height, width, in_channels, out_channels] -> (3,3,64,64)
> 
> Apply:-
> image_conv = tensorflow.nn.conv2d(in1, weight_kernel, strides=(1,1), padding='SAME', dilations=None)
> 
> Now this gives me shape of output of conv, same as input shape -> (1,56,56,64)
> 
> 
> 
> Over this I have added the biases that were extracted from the resnet model , Conv_2D_bias.shape (64,) - as shown in python and in TFlite image below.
> 
> this is done by simply doing image_conv + Conv_2D_bias
> 
> the output dimension remains the same - (1,56,56,64)
> 
> 
> Now I need to apply Relu activation from my side and compare with the existing relu output
> 
> The output from conv+bias is not visible from tflite model so cant compare.
> 
> Relu output is also different; and is mentioned as Fused activation function rather than just activation
> 
> My ques is how can I replicate this computation outside the model, as a second option.  The TFLite model of Resnet as in image was attached in .zip file in previous comment along with description, for reference.
> 
> ![image](https://user-images.githubusercontent.com/82860513/117398438-01dae580-af1c-11eb-8020-fb9ac0cf94f3.png)
> 

Thanks @thaink, I will relook into the flatbuffer topic, but this is something that will take time for me to do the setup for.

If you could let me know on this Relu query in python code I have quoted, this suits my experimentation topic better and I have already spend a few days progressing in it.

After applying the convolution and adding bias, how do I bring in the fused Relu correctly into picture so as to get the correct output as already in the model (given by the Identity_1). I find this method giving me good flexibility and better setup.

**Comment by manp-git** [comment]
Hi i generated the json file using some other schema.fbs link, found in one of the github links, as the one shared was giving error. I have not gone beyond that as I checked that it is rounding off q factor to 6 decimal places, also not so well versed with it. I recreated my own fucntion to do my job for now. Will get back if I try the flat compiler method. Thanks.

**Comment by dronefreak** [comment]
> Hi i generated the json file using some other schema.fbs link, found in one of the github links, as the one shared was giving error. I have not gone beyond that as I checked that it is rounding off q factor to 6 decimal places, also not so well versed with it. I recreated my own fucntion to do my job for now. Will get back if I try the flat compiler method. Thanks.

Hi @manp-git did you try the flatbuffers idea yet? Could you share your results with me please?

**Comment by manp-git** [comment]
> > Hi i generated the json file using some other schema.fbs link, found in one of the github links, as the one shared was giving error. I have not gone beyond that as I checked that it is rounding off q factor to 6 decimal places, also not so well versed with it. I recreated my own fucntion to do my job for now. Will get back if I try the flat compiler method. Thanks.
> 
> Hi @manp-git did you try the flatbuffers idea yet? Could you share your results with me please?

Hi @dronefreak , As I am not a part of this project now, so I couldn't try further on the flatbuffers and the issue is marked closed. Thanks for dropping by.
